File: postcodeSplitter.js
function splitPostcode(postcode) {
    if (postcode == null) { return null; }
    var fullPostcode = postcode.toUpperCase().trim();

    var inwardCode = '',
        outwardCode = '',
        isInward = false,
        characters = fullPostcode.split('');

    for (var index = 0; index < characters.length; index++) {
        var character = characters[index],
            nextCharacter = ' ';

        switch (index) {
            case 0:
                if (helpers.isLetter(character)) {
                    outwardCode += character;
                }
                break;
            case 1:
                if (helpers.isLetter(character)) {
                    outwardCode += character;
                }
                else if (helpers.isNumber(character)) {
                    outwardCode += character;
                    if (characters.length - 1 > index) {
                        nextCharacter = characters[index + 1];
                        if (helpers.isNumber(nextCharacter) && fullPostcode.length - 2 > index) {
                            nextCharacter = characters[index + 2];
                            isInward = helpers.isLetter(nextCharacter);
                        }
                    }
                }
                break;
            case 2:
                if (character == ' ') {
                    isInward = true;
                }
                else if (isInward) {
                    inwardCode += character;
                }
                else {
                    if (helpers.isLetter(character)) {
                        outwardCode += character;
                    }
                    else if (helpers.isNumber(character)) {
                        outwardCode += character;
                        if (characters.length - 1 > index) {
                            nextCharacter = characters[index + 1];
                            if (helpers.isNumber(nextCharacter) && characters.length - 2 > index) {
                                nextCharacter = characters[index + 2];
                                isInward = helpers.isLetter(nextCharacter);
                            }
                        }
                    }
                }
                break;
            case 3:
                if (isInward) {
                    if (helpers.isNumberOrLetter(character)) {
                        inwardCode += character;
                    }
                }
                else {
                    if (character == ' ') {
                        isInward = true;
                    }
                    else if (helpers.isLetter(character)) {
                        outwardCode += character;
                    }
                    else if (helpers.isNumber(character)) {
                        if (characters.length - 1 > index) {
                            nextCharacter = characters[index + 1];
                            if (helpers.isLetter(nextCharacter)) {
                                isInward = true;
                                inwardCode += character;
                            }
                            else {
                                outwardCode += character;
                            }
                        }
                        else {
                            outwardCode += character;
                        }
                    }
                }
                break;
            case 4:
                if (character == ' ') {
                    isInward = true;
                }
                else if (helpers.isNumberOrLetter(character)) {
                    inwardCode += character;
                }
                break;
            case 5:
                if (helpers.isNumberOrLetter(character)) {
                    inwardCode += character;
                }
                break;
            case 6:
                if (helpers.isNumberOrLetter(character)) {
                    inwardCode += character;
                }
                break;
            case 7:
                if (helpers.isNumberOrLetter(character)) {
                    inwardCode += character;
                }
                break;
        }
    }

    return { 'outwardCode': outwardCode, 'inwardCode': inwardCode };
}

File: PostcodeSplitter.cs
using System;

namespace Helpers
{
    public class Postcode
    {
        public string InwardCode { get; set; }
        public string OutwardCode { get; set; }
    }

    public static class PostcodeSplitter
    {
        public static Postcode Split(string postcode)
        {
            if (String.IsNullOrEmpty(postcode)) { return null; }

            var fullPostcode = postcode.ToUpper().Trim();
            var inwardCode = string.Empty;
            var outwardCode = string.Empty;
            var isInward = false;
            var characters = fullPostcode.ToCharArray();

            for (var i = 0; i < characters.Length; i++)
            {
                var character = characters[i];
                var nextCharacter = ' ';
                switch (i)
                {
                    case 0:
                        if (Char.IsLetter(character))
                        {
                            outwardCode += character;
                        }
                        break;
                    case 1:
                        if (Char.IsLetter(character))
                        {
                            outwardCode += character;
                        }
                        else if (Char.IsNumber(character))
                        {
                            outwardCode += character;
                            if (characters.Length - 1 > i)
                            {
                                nextCharacter = characters[i + 1];
                                if (Char.IsNumber(nextCharacter) && fullPostcode.Length - 2 > i)
                                {
                                    nextCharacter = characters[i + 2];
                                    isInward = Char.IsLetter(nextCharacter);
                                }
                            }
                        }
                        break;
                    case 2:
                        if (character == ' ')
                        {
                            isInward = true;
                        }
                        else if (isInward)
                        {
                            inwardCode += character;
                        }
                        else
                        {
                            if (Char.IsLetter(character))
                            {
                                outwardCode += character;
                            }
                            else if (Char.IsNumber(character))
                            {
                                outwardCode += character;
                                if (characters.Length - 1 > i)
                                {
                                    nextCharacter = characters[i + 1];
                                    if (Char.IsNumber(nextCharacter) && characters.Length - 2 > i)
                                    {
                                        nextCharacter = characters[i + 2];
                                        isInward = Char.IsLetter(nextCharacter);
                                    }
                                }
                            }
                        }
                        break;
                    case 3:
                        if (isInward)
                        {
                            if (Char.IsLetter(character) || Char.IsNumber(character))
                            {
                                inwardCode += character;
                            }
                        }
                        else
                        {
                            if (character == ' ')
                            {
                                isInward = true;
                            }
                            else if (Char.IsLetter(character))
                            {
                                outwardCode += character;
                            }
                            else if (Char.IsNumber(character))
                            {
                                if (characters.Length - 1 > i)
                                {
                                    nextCharacter = characters[i + 1];
                                    if (Char.IsLetter(nextCharacter))
                                    {
                                        isInward = true;
                                        inwardCode += character;
                                    }
                                    else
                                    {
                                        outwardCode += character;
                                    }
                                }
                                else
                                {
                                    outwardCode += character;
                                }
                            }
                        }
                        break;
                    case 4:
                        if (character == ' ')
                        {
                            isInward = true;
                        }
                        else if (Char.IsNumber(character) || Char.IsLetter(character))
                        {
                            inwardCode += character;
                        }
                        break;
                    case 5:
                        if (Char.IsLetter(character) || Char.IsNumber(character))
                        {
                            inwardCode += character;
                        }
                        break;
                    case 6:
                        if (Char.IsLetter(character) || Char.IsNumber(character))
                        {
                            inwardCode += character;
                        }
                        break;
                    case 7:
                        if (Char.IsLetter(character) || Char.IsNumber(character))
                        {
                            inwardCode += character;
                        }
                        break;
                }
            }
            return new Postcode() { InwardCode = inwardCode, OutwardCode = outwardCode };
        }
    }
}